engineering/documentation.md

```diff
@@ -0,0 +1,226 @@
+
+Give an overview of the project goal and its current state.
+Is this a demo, a library, or maybe some tool?
+Should it be used in production? Is too soon, or maybe too late because it's been deprecated?
+
+For tools, it is also worth including basic installation instructions.
+If the instructions are long they should be part of a different document linked from this one.
+
+This document also serves as the index for all other documentation.
+`README.md` can also be found in folders other than the root of the repository.
+In that case it should document what the subtree below the folder contains.
+This is useful when you have libraries and binaries in a single repository.
+
+TL;DR:
+- Audience: Everyone.
+- Content: Description of the project and its current state. Links to other resources.
+- Example: https://github.com/twbs/bootstrap/blob/v4-dev/README.md
+
+### CONTRIBUTING.md
+
+This document describes what contributors should know in order to contribute to the project.
+
+This document should contain all the details required for anyone to send a contribution.
+It often contains instructions on how to build the project from source code and validate the
+source code by running unit and integration tests.
+
+This is also the place where style guides and other code conventions and requirements
+should be documented or at least referenced.
+Examples of these include code coverage requirements, adopted unit testing frameworks,
+and documentation conventions.
+
+A very good example of `CONTRIBUTING.md` is the one from [angular.js](https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md).
+
+TL;DR:
+- Audience: Contributors.
+- Content: Steps necessary to modify the project, verify the changes, and contribute back to the project.
+- Example: https://github.com/angular/angular.js/blob/master/README.md
+
+### Reference material
+
+This is the documentation that users will reguarly consult while using your project.
+The actual form depends on whether the project is a library or a tool.
+
+For libraries, this document tends to be autogenerated from source code.
+Tools like godoc, javadoc, etc. provide an easy way to do this.
+For tools, the document will be often hand written, think about `man` pages.
+
+In both cases, this kinds of document tend to grow quite fast as they provide
+detailed information on each feature and the feature's options.
+They do not require a sequential structure. Users often search and read single section.
+Therefore it's recommended to have a good table of contents and search features when possible.
+
+TL;DR:
+- Audience: Users of the project.
+- Content: List of all of the features of the project with full information on arguments, flags, and others.
+- Example: https://docs.angularjs.org/api
+
+### Release Notes
+
+Every new release of your project should be versioned following semantic versioning principles.
+Each new release should come with release notes. These notes are not a list of every single
```

Comment:
True, but sometime expensive.

Reply:
I think it's not necessarily expensive. The notes are *not* exhaustive. Basically what's in the release that I might care about.